src/kinship/hierarchy-creator.ts

```diff
@@ -0,0 +1,303 @@
+      spouseSiblings: filter.spouseSiblings ? this.indiIdsToFamAsSpouseNodes(spouseSiblingsIds, parentNode, LinkType.SpouseSiblings) : [],
+      children:       filter.children       ? this.indiIdsToFamAsSpouseNodes(childrenIds, parentNode, LinkType.Children)             : [],
+    });
+  }
+
+  private childNodesForIndi(indi: Indi, parentNode: TreeNode, filter: HierarchyFilter): ChildNodes {
+    const [indiParentsFamsIds, indiSiblingsIds] = this.getParentsAndSiblings(indi);
+    return new ChildNodes({
+      indiParents:    filter.indiParents  ? this.famAsSpouseIdsToNodes(indiParentsFamsIds, parentNode, LinkType.IndiParents)   : [],
+      indiSiblings:   filter.indiSiblings ? this.indiIdsToFamAsSpouseNodes(indiSiblingsIds, parentNode, LinkType.IndiSiblings) : [],
+    });
+  }
+
+  private areParentsAndSiblingsPresent(indiId: string): [boolean, boolean] {
+    const indi = this.data.getIndi(indiId);
+    const famc = this.data.getFam(indi && indi.getFamilyAsChild());
+    if (!famc) return [false, false];
+    return [
+      !!(famc.getFather() || famc.getMother()),
+      famc.getChildren().length > 1,
+    ];
+  }
+
+  private getParentsAndSiblings(indi: Indi): [string[], string[]] {
+    const indiFamcId = indi ? indi.getFamilyAsChild() : null;
+    const indiFamc = this.data.getFam(indiFamcId);
+    if (!indiFamc) return [[], []];
+
+    const [father, mother] = [indiFamc.getFather(), indiFamc.getMother()].map(id => this.data.getIndi(id));
+    const parentFamsIds = [].concat(
+      father ? father.getFamiliesAsSpouse() : [],
+      mother ? mother.getFamiliesAsSpouse() : []
+    ).filter(id => id !== indiFamcId);
+    parentFamsIds.unshift(indiFamcId);
+
+    const siblingsIds = Array.from(indiFamc.getChildren());
+    siblingsIds.splice(siblingsIds.indexOf(indi.getId()), 1);  // Remove indi from indi's siblings
+
+    return [parentFamsIds, siblingsIds];
+  }
+
+  private indiIdsToFamAsSpouseNodes(indiIds: string[], parentNode: TreeNode, childNodeType: LinkType): TreeNode[] {
+    return indiIds.flatMap(id => this.indiIdToFamAsSpouseNodes(id, parentNode, childNodeType));
+  }
+
+  private indiIdToFamAsSpouseNodes(indiId: string, parentNode: TreeNode, childNodeType: LinkType): TreeNode[] {
+    if (this.isChildNodeTypeForbidden(childNodeType, parentNode)) return [];
+    const famsIds = this.data.getIndi(indiId).getFamiliesAsSpouse();
+    if (!famsIds.length) {
+      const node = this.idToNode(EntryId.indi(indiId), parentNode, childNodeType);
+      return node ? [node] : [];
+    }
+
+    const famsNodes: TreeNode[] = famsIds.map(id => {
+      return {
+        id,
+        indi: {id: indiId},
+        family: {id},
+        parentNode,
+        linkFromParentType: childNodeType,
+        childNodes: null,
+        linkStubs: null,
+        duplicateOf: this.queuedNodesById.get(id),
+      };
+    });
+    famsNodes.forEach((node, i) => {
+      if (i !== 0) node.primaryMarriage = famsNodes[0];
```

Comment:
Isn't this equivalent to
`if (i !== 0) node.primaryMarriage = node`;
?

Reply:
No. `node` is equal to `famsNodes[0]` only when `i === 0`, and here we assign `famsNodes[0]` to `node.primaryMarriage` for every node in `famsNodes` *except* the first one (when `i  !== 0`).